scala/src/main/org/apache/spark/streaming/api/kafka/KafkaOffsetsFetcher.scala

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) Microsoft. All rights reserved.
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+
+package org.apache.spark.streaming.kafka
+
+import scala.annotation.tailrec
+
+import java.util.UUID
+import java.util.concurrent.{LinkedBlockingQueue, TimeUnit}
+
+import kafka.common.TopicAndPartition
+
+import org.apache.spark.Logging
+import org.apache.spark.streaming.StreamingContext
+import org.apache.spark.streaming.kafka.KafkaCluster.LeaderOffset
+import org.apache.spark.util.ThreadUtils
+
+case class KafkaOffsetsFetcherStateSnapshot(
+     hasFetchedAllPartitions: Boolean,
+     hasFetchedAllFromOffsets: Boolean,
+     topicAndPartitions: Set[TopicAndPartition],
+     currentOffsets: Map[TopicAndPartition, Long])
+
+case class KafkaOffsetsFetchResult(
+    fromOffsets: Map[TopicAndPartition, Long],
+    untilOffsets: Map[TopicAndPartition, LeaderOffset],
+    fetcherStateSnapshot: KafkaOffsetsFetcherStateSnapshot)
+
+/**
+ * reading metadata of multiple topics from across multiple data centers takes long time to
+ * complete, which impacts DStream performance and causes UI steaming tab not responsive due to
+ * mutex held by DStream, so a separate thread is introduced to refresh metadata (current offsets)
+ * asynchronously at below interval. This unblocks DStream in above described situation but not
+ * quite in sync with batch timestamp, which is OK since batches are still generated at the same
+ * interval
+ */
+class KafkaOffsetsFetcher(
```

Comment:
I will have a try. But the UNIT test of this class involves Kafka and ZooKeeper and may be difficult to do.